### PR TITLE
Feature/district level export page

### PR DESCRIPTION
**Pull request by esotrope**
Adds placeholder for new "Customized Export" page.
Adds inbound links to page as well.

Like "My Reports," currently the page is hosted at /custom-export under the the root path though it it can be linked to from the home page and each test result page.

Also, if I merge I'm tempted to take the links out until there is actually something on the page to test. thoughts?

screen shot 2017-09-25 at 2 03 43 pm

Homepage link alone

screen shot 2017-09-25 at 1 34 40 pm

Homepage link beside admin link

screen shot 2017-09-25 at 1 34 27 pm

Result page link

screen shot 2017-09-25 at 3 53 16 pm

### CODE REVIEW

File: webapp/src/main/webapp/src/app/custom-export/custom-export.module.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+import { CommonModule } from "../shared/common.module";
+import { CustomExportComponent } from "./custom-export.component";
+import { UserModule } from "../user/user.module";
+
+@NgModule({
+  declarations: [
+    CustomExportComponent
+  ],
+  imports: [
+    BrowserAnimationsModule,
```
Comment (jtreuting) [inline webapp/src/main/webapp/src/app/custom-export/custom-export.module.ts]: Do we need animations?

File: webapp/src/main/webapp/src/app/groups/groups.module.ts [hunk 2]
```diff
@@ -28,6 +31,7 @@ import { PopoverModule } from "ngx-bootstrap";
     SharedModule,
     PopoverModule.forRoot(),
     ReportModule,
+    CustomExportModule,
```
Comment (jtreuting) [inline webapp/src/main/webapp/src/app/groups/groups.module.ts]: Is the module used on this page or does it just link to it?

File: webapp/src/main/webapp/src/app/custom-export/custom-export.module.ts (hunk 1 above)
Comment (esotrope) [reply webapp/src/main/webapp/src/app/custom-export/custom-export.module.ts]: Shouldn't no - I'll try and remove this.

File: webapp/src/main/webapp/src/app/groups/groups.module.ts (hunk 2 above)
Comment (esotrope) [reply webapp/src/main/webapp/src/app/groups/groups.module.ts]: This is leftover I think. Originally it was importing a common link component for reuse.

**Comment by esotrope** [comment]
@bgsavage @wtritch @jtreuting @malaffoon @aaronsleeper 

Made the following functional additions:

1. Search select / typeahead

screen shot 2017-09-28 at 10 18 00 am

screen shot 2017-09-28 at 10 17 36 am

2. Org tree view

(districts and schools only)

screen shot 2017-09-28 at 10 16 54 am

(all org types)

screen shot 2017-09-28 at 10 16 08 am

Some concerns about this view and performance:

When clicking show/all and removing an element in Chrome I observed the following rough response times for the given amount of organizations:
  320 < .5s (responsive)
  360 < .5s (responsive)
  400 ~ .5s
  500 ~ 1s
  800 ~ 1.5s
  1000 ~ 2s (LA school District)
  2000 ~ 2s
  4000 ~ 3s
  8000 ~ 6s
  16000 page becomes completely unresponsive

Here is the script/dom rendering time distribution (2/5 script, 3/5 dom)

screen shot 2017-09-28 at 3 54 34 pm

File: webapp/src/main/webapp/src/app/custom-export/custom-export.component.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+          label: organization.name,
+          group: this.translate.instant(`labels.custom-export.form.organization.type.${OrganizationType[ organization.type ]}`),
+          value: organization
+        });
+
+      // recompute the organizations in the tree
+      this._tree = this.mapper
+        .organizationTree(value)
+        .sort(this._comparator);
+    }
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/custom-export/custom-export.component.ts]: Basically the strategy here is to regenerate the search options list and org tree every time the selection is changed.
This is not ideal for performance but is the most straight forward and bug proof way to develop this for a first pass.

File: webapp/src/main/webapp/src/app/shared/form/search-select.ts [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+             [(ngModel)]="search"
+             (ngModelChange)="onSearch($event.value)"
+             [typeahead]="options"
+             typeaheadOptionField="label"
+             typeaheadGroupField="group"
+             (typeaheadOnSelect)="onSelect($event.item)"
+             [placeholder]="searchPlaceholder">
+  `
+})
+export class SearchSelect {
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/shared/form/search-select.ts]: This component is a generalized version of the homepage school dropdown typeahead and can be used in its place

File: webapp/src/main/webapp/src/app/custom-export/organization/organization.service.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+            id: school.id,
+            name: school.name,
+            schoolId: school.id,
+            schoolGroupId: school.schoolGroupId,
+            schoolGroupName: schoolGroupNamesById.get(school.schoolGroupId),
+            districtId: school.districtId,
+            districtName: districtNamesById.get(school.districtId),
+            districtGroupId: school.districtGroupId,
+            districtGroupName: districtGroupNamesById.get(school.districtGroupId)
+          });
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/custom-export/organization/organization.service.ts]: This method flattens the responses into a collection of schools containing all ancestry id/name pairs.
This is later used by the mapper to derive the select options and org tree model.

File: webapp/src/main/webapp/src/app/custom-export/organization/organization.mapper.ts [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    return organizations;
+  }
+
+  organizationTree(schools: FlatSchool[]): Tree<Organization> {
+    let root = new Tree<Organization>();
+    schools.forEach(school => root
+      .getOrCreate(x => x.id === school.districtGroupId, this.districtGroup(school))
+      .getOrCreate(x => x.id === school.districtId, this.district(school))
+      .getOrCreate(x => x.id === school.schoolGroupId, this.schoolGroup(school))
+      .create(this.school(school))
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/custom-export/organization/organization.mapper.ts]: known-depth iterative solution to generating the tree model.

**Comment by esotrope** [comment]
Performance is addressed by using flyweight organization objects and not abusing the heap by regenerating them every render.
Also, district groups are no longer displayed and only one district is allowed to be selected at a time.
Show all link is now not displayed when a user is entitled to more than one district.

**Pull request merged**